### PR TITLE
Parallel Replica Exchange

### PR DESCRIPTION
**Pull request by SamWitty**
Added Julia multiprocessing and more extensive testing.

### CODE REVIEW

**Review by zenna — CHANGES_REQUESTED** [review]
Hey Sam,

Some edits (and I realize most of these are cleaning up my mess)
- Help of `re!` should state that it mutates `states`
- Help of `re` should:
  - Using @Ref to refer to `re!` (search @Ref here https://docs.julialang.org/en/v1/manual/documentation/_)
  - Say that it does not mutate `states` rather than just say not mutating

- Delete all the commented questions I have at the bottom
- Is `re_all` functional? If so, give it a docstring and tests, if not, delete it.
- Add preconditions and postconditions using Spec.jl.  Some examples of adding specs can be found here https://github.com/zenna/Omega.jl/blob/master/OmegaCore/src/util/namedtuple.jl

- It should be possible for me to cd into ReplicaExchange, and do `julia --project` then `] test`
Currently this is broken because ReplicaExchange doesn't have the right test dependencies
You'll need to add `extras`  See Project.toml for OmegaCore as a guide

One slightly tricky thing there is that the tests of ReplicaExchange depend on OmegaMH which is not registered.  One way to handle this is to add `Pkg.develop(....  )`
See `OmegaMH/test/runtests.jl` for an example of this

Finally `runtests` should also run the notebook, so I'd recommend moving most of the current content of `runtests` into a new file, and having `runtests` simply include the notebook

Finally finally I'm actually running the tests now and they haven't finished.  I don't yet know if this is because of a bug or because you are doing an extremely compute-heavy computation.  These tests are going to run everytime we commit to master and so they shouldn't be more compute-heavy than necessary

(I have also not yet reviewed the notebook and will add comments on that in a moment)

**Comment by SamWitty** [comment]
Thanks for the feedback/requests.

Re: tests being slow, I think this is something we'll need to think a bit more about when testing approximate inference methods. The current test suite includes some pretty forgiving assessments of inference "correctness": (i) it only tests a single inference run for each of the two target densities, as opposed to the many many runs required by methods like [SBC](https://arxiv.org/pdf/1804.06788.pdf) and (ii) it only requires that a handful of summary statistics (mean, variance, proportion of samples in positive quadrant) approximately match the truth. Even so, I'm finding it very difficult to get the tests to pass even these permissive assessments with much fewer iterations than are currently in the test suite. 

As far as I see it we have a few different options:
1. Remove "inference correctness" tests altogether, and simply test whether the inference method returns the right kinds of objects with the right kinds of inputs.
2. Keep the test suite as-is, or even it slower by making the tests more principled.
3. Stage the tests so that "inference correctness" tests run less often and standard "does it run" tests run on every commit to master.

My preference is 3, but I'm not sure how involved it is to change Omega's CI/CD workflows to make that happen.

Thoughts?

File: ReplicaExchange/src/recore.jl [hunk 1]
```diff
@@ -92,103 +99,194 @@ function re!(rng,
...
     end
     swap_contexts!(rng, logenergys, states)
   end
   samples
 end
-# @pre length(logenergys) == length(states) "Must have one density per initial state"
 
-"Non-mutating re!"
+@pre re!(rng, logenergys, samples_per_swap, num_swaps, states, samples, sim_chain_keep_n) = num_swaps > 0 "There must be at least 1 swap"
+@pre re!(rng, logenergys, samples_per_swap, num_swaps, states, samples, sim_chain_keep_n) = all([isa(logenergy, Function) for logenergy in logenergys]) "logenergys is a Function"
```
Comment (zenna) [inline ReplicaExchange/src/recore.jl:132]: `isa(logenergy, Function)` does not need to be true (it's overly restrictive)

Comment (zenna) [reply ReplicaExchange/src/recore.jl:132]: a logenergy need only be some object on which logenergy(state) is defined, which may not be a subtype of `Function`

File: ReplicaExchange/src/recore.jl [hunk 2]
```diff
@@ -92,103 +99,194 @@ function re!(rng,
...
+@pre re!(rng, logenergys, samples_per_swap, num_swaps, states, samples, sim_chain_keep_n) = isa(sim_chain_keep_n, Function) "sim_chain_keep_n is a Function"
+@pre re!(rng, logenergys, samples_per_swap, num_swaps, states, samples, sim_chain_keep_n) = length(logenergys) == length(states) "Must have one density per initial state"
+# usage of @cap and @ret in README of Spec.jl throws an error.
+# @post re!(rng, logenergys, samples_per_swap, num_swaps, states, samples, sim_chain_keep_n, sim_chain_keep_last, swap_contexts!) = (@cap(x), @ret) "Result is sorted version of input"
+
```
Comment (zenna) [inline ReplicaExchange/src/recore.jl:137]: This is __ret__ - Look in https://github.com/zenna/Spec.jl/blob/master/test/runtests.jl

File: ReplicaExchange/src/recore.jl [hunk 3]
```diff
@@ -92,103 +99,194 @@ function re!(rng,
...
+          # If not ground state just return last sample
+          @inbounds states[i] = sim_chain_keep_last(rng, logenergys[i], states[i], samples_per_swap, i)
+        end
       end
     end
     swap_contexts!(rng, logenergys, states)
   end
   samples
 end
-# @pre length(logenergys) == length(states) "Must have one density per initial state"
```
Comment (zenna) [inline ReplicaExchange/src/recore.jl:130]: This is missing a precondition on `samples`- see line 119

File: ReplicaExchange/src/recore.jl [hunk 4]
```diff
@@ -92,103 +99,194 @@ function re!(rng,
...
+  a user-specified transition kernel (`sim_chain_keep_n`), and then subsequently proposing 
+  a special transition kernel that swaps the position of parallel chains.
+- REMCMC is agnostic to the user-specified transition kernel, as long as its stationary distribution
+  is equal to the distribution corresponding to its respective input density. E.g. this permits HMC or RW-MCMC.
+- A common use-case for REMCMC is when the target distribution is non-smooth and we would expect
+  a single MCMC transition kernel to become stuck near a local optima. Then, the collection of distributions
+  are induced by annealing the target distribution with a sequence of increasing temperatures.
+- In this implementation we assume that the user is only interested in samples from the single target distribution
+  and that the same `sim_chain_keep_n` is used for each of the `n` distributions.
+- Note: re does not mutate `states`.
```
Comment (zenna) [inline ReplicaExchange/src/recore.jl:165]: Don't think you need to repeat yourself (reference and maybe example are sufficient), see help on `sort!!` and `sort`

File: ReplicaExchange/Project.toml [hunk 5]
```diff
@@ -11,3 +11,17 @@ Spec = "b8ccf107-3a88-5e0f-823b-b838c6a0f327"
...
+
+[extras]
+Distributions = "31c24e10-a181-5473-b8eb-7969acd0382f"
+InteractiveUtils = "b77e0a4c-d291-57a0-90e8-8db25a27a240"
+Markdown = "d6f4376e-aef5-505a-96c1-9c027394607a"
+Pkg = "44cfe95a-1eb2-52ea-b672-e2afdf69b78f"
+PlutoUI = "7f904dfe-b85e-4ff6-b463-dae2292396a8"
+Plots = "91a5bcdd-55d7-5caf-9e0b-520d859cae80"
+Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
+Revise = "295af30f-e4ad-537b-8983-00126c2a3abe"
```
Comment (zenna) [inline ReplicaExchange/Project.toml:23]: You don't need Revise in these dependencies

Comment (zenna) [reply ReplicaExchange/Project.toml:23]: Add it to your local config

File: ReplicaExchange/test/tests.jl [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+using Distributions
+using Spec
+import Random
+
+# Gaussian drift proposal
+function propose_and_logratio(rng, state)
+	prop_dist = MvNormal(state, [0.1, 0.1])
+	sample = rand(rng, prop_dist)
+	(sample, 0)
+end
```
Comment (zenna) [inline ReplicaExchange/test/tests.jl:17]: Omega.jl uses two space indentation - you can use the julia formatter/linter to automate this

File: ReplicaExchange/test/tests.jl [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+  # # re approximately matches the mean and variance
+  # @test mean(samples) ≈ mean(dist) atol = 0.3
+  # @test var(samples) ≈ var(dist) atol = 0.3
+
+  # # re approximately matches the proportion in upper right quadrant (easy proxy for capturing multimodality in this special case)
+  # # Distributions.jl doesn't have a cdf method, so we'll just take a Monte Carlo estimate.
+  # proportion = mean(sum(hcat(samples...) .> 0, dims = 1) .> 0)
+  # true_samples = rand(rng, dist, num_samples)
+  # true_proportion = mean(sum(true_samples .> 0, dims = 1) .> 0)
+  # @test proportion ≈ true_proportion atol = 0.1
```
Comment (zenna) [inline ReplicaExchange/test/tests.jl:87]: Why are these commented?

**Review by zenna — CHANGES_REQUESTED** [review]
Sorry for delay. Almost there

Other than comments

- Remove Revise from notebook
- Make it work locally by adding the following lines
- `import Pkg`
- Pkg.activate(Base.current_project())`

reactivate or delete (remember, version control remembers everything) commented code. 
If some of those are concerning broader questions about approximate testing, let's save that for another PR

File: ReplicaExchange/test/tests.jl (hunk 7 above)
Comment (SamWitty) [reply ReplicaExchange/test/tests.jl:87]: These were commented out to reflect my uncertainty in whether to include tests of approximate inference.

I will delete them for now, and then (as you mention) we can revisit approximate inference tests in a separate PR.

File: ReplicaExchange/Project.toml (hunk 5 above)
Comment (SamWitty) [reply ReplicaExchange/Project.toml:23]: Revise and Plots are used in the notebook. Would you like me to remove them?

**Comment by zenna** [comment]
Nice work

**Pull request merged**